controllers/config_helper.go

```diff
@@ -0,0 +1,192 @@
+package controllers
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/dynamic"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var errConfigCrdNotReady = errors.New("Config CRD is not ready")
+
+// Default config data
+var configDefaultResource = []byte(`
+apiVersion: config.gatekeeper.sh/v1alpha1
+kind: Config
+metadata:
+  name: config
+spec:
+  match:
+  - excludedNamespaces: ["kube-*", "multicluster-engine", "hypershift", "hive", "rhacs-operator", "open-cluster-*"] 
+    processes: ["audit", "webhook", "sync"]
+`)
+
+func createConfigFromYaml(ctx context.Context, c client.Client, namespace string) error {
+	config := &v1alpha1.Config{}
+
+	err := yaml.Unmarshal(configDefaultResource, config)
+	if err != nil {
+		return err
+	}
+
+	config.ObjectMeta = metav1.ObjectMeta{
+		Name:      "config",
+		Namespace: namespace,
+	}
+
+	err = c.Create(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkConfigHasExemptNs(ctx context.Context,
+	c client.Client, existingConfig v1alpha1.Config, log logr.Logger,
+) error {
+	configDefault := &v1alpha1.Config{}
+
+	err := yaml.Unmarshal(configDefaultResource, configDefault)
+	if err != nil {
+		return err
+	}
+
+	// Check only ExcludedNamespaces not processes
+	// Users can edit processes
+
+	for _, match := range existingConfig.Spec.Match {
+		if reflect.DeepEqual(match,
+			configDefault.Spec.Match[0]) {
+			log.Info("Default excludedNamespaces exist already")
+
+			return nil
+		}
+	}
+
+	existingConfig.Spec.Match = append(existingConfig.Spec.Match, configDefault.Spec.Match[0])
+
+	err = c.Update(ctx, &existingConfig, &client.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	log.Info("Update config because Default excludedNamespaces were missed")
+
+	return nil
+}
+
+func (r *GatekeeperReconciler) handleConfigController(mainCtx context.Context) error {
+	isCRDReady, err := checkConfigCrdAvailable(mainCtx, r.DynamicClient)
+	if err != nil {
+		return err
+	}
+
+	if !isCRDReady {
+		return errCpsCrdNotReady
+	}
+
+	if r.isConfigCtrlRunning {
+		return nil
+	}
+
+	configMgr, err := ctrl.NewManager(r.KubeConfig, ctrl.Options{
```

Comment:
Could you please limit the watch/cache to just the name `config` since I believe that's the only one you look at?
Here is an example of how you do that:
https://github.com/open-cluster-management-io/governance-policy-propagator/blob/f81d2267e2794bcd8240e358176bad7433d80748/main.go#L232-L234

Reply:
I have question does this not affect when we `update` config?

Reply:
Yeah, this doesn't limit which fields are cached or returned, it just limits the watch API request to just the `Config` object called `config`.

Reply:
@mprahl  I tested. `Update` needs everything otherwise, it doesn't update properly. I limited only spec.match after update it lost `spec.sync`, `spec.violation` and so on. Also `Update`  needs metadata too

Reply:
I added 
```
		Cache: cacheRuntime.Options{
			ByObject: map[client.Object]cacheRuntime.ByObject{
				&v1alpha1.Config{}: {
					Namespaces: map[string]cacheRuntime.Config{
						r.Namespace: {
							FieldSelector: fields.SelectorFromSet(fields.Set{"metadata.name": "config"}),
						},
					},
				},
			},
```